### ISSUE TITLE
[BUG] bundledeployment rancher-monitoring always has status "modified"

### ISSUE THREAD

**Issue by w13915984028** [harvester/harvester#2152]
**Describe the bug**

For a successfully installed Harvester cluster, it runs well, but the bundledeployment` rancher-monitoring` always has status `"modified"`  with `"mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified"`.

```
$ kubectl get bundledeployment -A 

NAMESPACE                                NAME                                          DEPLOYED   MONITORED   STATUS
..
cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring                        True       True        mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified {"webhooks":[{"admissionReviewVersions":["v1","v1beta1"],"clientConfig":{"caBundle":"LS..=","service":{"name":"rancher-monitoring-operator","namespace":"cattle-monitoring-system","path":"/admission-prometheusrules/mutate","port":443}},"failurePolicy":"Ignore","matchPolicy":"Equivalent","name":"prometheusrulemutate.monitoring.coreos.com","namespaceSelector":{},"objectSelector":{},"reinvocationPolicy":"Never","rules":[{"apiGroups":["monitoring.coreos.com"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["prometheusrules"],"scope":"*"}],"sideEffects":"None","timeoutSeconds":10}]}
cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring-crd                    True       True  
```

**To Reproduce**
```
1. Install Harvester cluster (one node or multi nodes)
2. When the system UI is ready, execute "kubectl get bundledeployment -A ", get the related status.
3. Meanwhile, continuous retrying in the log of fleet-agent.
```

**Expected behavior**

System should be steady, no such warnings.

**Support bundle**

**Environment:**
 - Harvester ISO version:  v1.0.1
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): KVM based.

**Additional context**
Other related issues:
#1531 
#2013

**Comment by w13915984028** [comment]
Analysis:

The rancher-monitoring chart, which is used by Harvester, is at  https://github.com/rancher/charts/tree/dev-v2.6/charts/rancher-monitoring/100.1.0%2Bup19.0.3

1. Values in values.yaml
the expected value of `admissionWebhooks.failurePolicy` is `Failue`
```
prometheusOperator.admissionWebhooks.failurePolicy: Fail
prometheusOperator.admissionWebhooks.patch.enabled: true
```



2. When install/upgrade, the `admissionWebhooks.failurePolicy` value is set to: `Ignore` when `patch.enabled` is true, seems for skipping any potential patch failure
templates/prometheus-operator/admission-webhooks/mutatingWebhookConfiguration.yaml 
```
webhooks:
  - name: prometheusrulemutate.monitoring.coreos.com
    {{- if .Values.prometheusOperator.admissionWebhooks.patch.enabled  }}
    failurePolicy: Ignore
    {{- else }}
    failurePolicy: {{ .Values.prometheusOperator.admissionWebhooks.failurePolicy }}
    {{- end }}

```

3. A post-install/post-upgrade hook is used to set back the value of `admissionWebhooks.failurePolicy` as per defined in values.yaml
templates/prometheus-operator/admission-webhooks/job-patch/job-patchWebhook.yaml 

```
kind: Job
  annotations:
    "helm.sh/hook": post-install,post-upgrade
...
    spec:
..
      containers:
        - name: patch
..
          args:
            - patch
            - --webhook-name={{ template "kube-prometheus-stack.fullname" . }}-admission
            - --namespace={{ template "kube-prometheus-stack.namespace" . }}
            - --secret-name={{ template "kube-prometheus-stack.fullname" . }}-admission
            - --patch-failure-policy={{ .Values.prometheusOperator.admissionWebhooks.failurePolicy }}


```
4. The stable running value of `admissionWebhooks.failurePolicy`
```
$ kubectl get mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission -o JSON 

	webhooks
...
            "failurePolicy": "Fail",
```



5. The 'bundledeployment' mcc-rancher-monitoring always complains "modified", further, the fleet-agent controller reconciles it continuously

```
NAMESPACE                                NAME                                          DEPLOYED   MONITORED   STATUS
cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring                        True       True        mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified {"webhooks":[{"admissionReviewVersions":["v1","v1beta1"],"clientConfig":{"caBundle":"LS...LQo=","service":{"name":"rancher-monitoring-operator","namespace":"cattle-monitoring-system","path":"/admission-prometheusrules/mutate","port":443}},"failurePolicy":"Ignore","matchPolicy":"Equivalent","name":"prometheusrulemutate.monitoring.coreos.com","namespaceSelector":{},"objectSelector":{},"reinvocationPolicy":"Never","rules":[{"apiGroups":["monitoring.coreos.com"],"apiVersions":["*"],"operations":["CREATE","UPDATE"],"resources":["prometheusrules"],"scope":"*"}],"sideEffects":"None","timeoutSeconds":10}]}

note: "failurePolicy":"Ignore"

```


Per Helm documents:

https://helm.sh/docs/topics/charts_hooks/
```
pre-install Executes after templates are rendered, but before any resources are created in Kubernetes
post-install Executes after all resources are loaded into Kubernetes
...
pre-upgrade  Executes on an upgrade request after templates are rendered, but before any resources are updated
post-upgrade Executes on an upgrade request after all resources have been upgraded
...
```


Even though it is not explictly wroten, but as said 'post-upgrade Executes on an upgrade request **after all resources have been upgraded**' ,  **using "post-install" or "post-upgrade" hook to change kubernetes resources seems not a good idea, it will cause HELM to treat the related chart as not ready**.

**Comment by w13915984028** [comment]
Another post-install, post-upgrade hook causes similar problem.

https://github.com/harvester/harvester/issues/2088

**Comment by w13915984028** [comment]
@[aiyengar2](https://github.com/aiyengar2)
Per our previous discussion:
https://github.com/harvester/harvester/issues/1893#issuecomment-1033072423

It looks such a "writing" post-upgrade/post-install hook is the root cause.

Do you have any idea how to solve it? thanks.

**Comment by w13915984028** [comment]
If HELM is used manually, then the post-install/upgrade hook seems OK;   but when embedded in fleet-agent, it is not feasible now,  fleet-agent does a check before&after HELM install/upgrade, complains post-* hook writing.

e.g.

```
harv41:/var/lib/rancher/rancherd/plan/work/20220419-130925-applied.plan/_4 #

./helm history rancher-monitoring --namespace cattle-monitoring-system
REVISION	UPDATED                 	STATUS    	CHART                              	APP VERSION	DESCRIPTION     
1       	Tue Apr 19 13:11:58 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Install complete
2       	Tue Apr 19 13:14:59 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
3       	Tue Apr 19 13:21:52 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
4       	Tue Apr 19 14:59:02 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
5       	Tue Apr 19 21:00:46 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
6       	Tue Apr 19 21:01:30 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
7       	Tue Apr 19 21:11:09 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete

8       	Wed Apr 20 08:38:08 2022	superseded	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
9       	Wed Apr 20 08:47:49 2022	deployed  	rancher-monitoring-100.1.0+up19.0.3	0.50.0     	Upgrade complete
```
![image](https://user-images.githubusercontent.com/31133476/164213007-0369b25e-bf6b-4dcb-85bb-678a0ee6e0ab.png)

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.0.2-rc1`. Close this issue.

### Result
* The bundledeployment `rancher-monitoring` now don't have status "modified" with "mutatingwebhookconfiguration.admissionregistration.k8s.io rancher-monitoring-admission modified"

  ```
  hpd8s7:~ # kubectl get bundledeployment -A
  NAMESPACE                                NAME                                          DEPLOYED   MONITORED   STATUS
  cluster-fleet-local-local-1a3d67d0a899   fleet-agent-local                             True       True        
  cluster-fleet-local-local-1a3d67d0a899   local-managed-system-agent                    True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-harvester                                 True       True        configmap.v1 longhorn-system/longhorn-storageclass missing
  cluster-fleet-local-local-1a3d67d0a899   mcc-harvester-crd                             True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-local-managed-system-upgrade-controller   True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring                        True       True        
  cluster-fleet-local-local-1a3d67d0a899   mcc-rancher-monitoring-crd                    True       True
  ```
* Re-trying the fleet-agent logs looks fine
  ![image](https://user-images.githubusercontent.com/29251855/165931544-dd02ab8c-9ee1-4dc1-a139-7cb7f32e9977.png)


### Test Information
* Test Environment: 3 nodes harvester on provo bare machine
* Harvester version: v1.0.2-rc1

### Verify Steps

1. Prepare the 3 nodes Harvester
2. ssh to vip node
3. When the system UI is ready, execute `kubectl get bundledeployment -A `, get the related status.
4. Open another terminal and ssh to vip node
5. Meanwhile, continuous retrying in the log of fleet-agent.

### Support Bundle
[supportbundle_bea46fd6-ab74-4853-9786-d2a29898c248_2022-04-29T11-21-23Z.zip](https://github.com/harvester/harvester/files/8590395/supportbundle_bea46fd6-ab74-4853-9786-d2a29898c248_2022-04-29T11-21-23Z.zip)
